### PR TITLE
Feature/openid

### PR DESCRIPTION
**Pull request by ThaerAlAjlouni**
This PR adds OpenId to Orchard. See #7419

### CODE REVIEW

File: src/Orchard.Web/Modules/Orchard.OpenId/Views/EditorTemplates/Parts.AzureActiveDirectorySettings.cshtml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+             Example: service1=https://yoursite.onmicrosoft.com/some-guid-for-service1")
+    </span>
+</fieldset>
+<fieldset>
+    @Html.LabelFor(m => m.AppKey, T("App Key"))
+    @Html.TextBoxFor(m => m.AppKey, new { @class = "text large" })
+</fieldset>
+<fieldset>
+    @Html.CheckBoxFor(m => m.BearerAuthEnabled)
+    @Html.LabelFor(m => m.BearerAuthEnabled, T("Enable Bearer Token Authentication").ToString(), new { @class = "forcheckbox" })
```
Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.OpenId/Views/EditorTemplates/Parts.AzureActiveDirectorySettings.cshtml]: Don't use `LabelFor` and `T.ToString` at the same time or there will be a double encoding. Use a `<label>` tag instead. Everywhere in this PR.

File: src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        }
+
+        public void Load() {
+            _sessionLock.EnterReadLock();
+            try
+            {
+                if (InMemoryTokenCache.ContainsKey(_cacheId))
+                    Deserialize(InMemoryTokenCache.Where(x => x.Key == _cacheId).First().Value);
+            }
+            catch (Exception) { /* TODO: Log the error */ }
```
Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs]: Please implement the TODOs

File: src/Orchard.Web/Modules/Orchard.OpenId/Views/EditorTemplates/Parts.AzureActiveDirectorySettings.cshtml (hunk 1 above)
Comment (ThaerAlAjlouni) [inline src/Orchard.Web/Modules/Orchard.OpenId/Views/EditorTemplates/Parts.AzureActiveDirectorySettings.cshtml]: Done

File: src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs (hunk 2 above)
Comment (ThaerAlAjlouni) [inline src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs]: Done

File: src/Orchard.Web/Modules/Orchard.OpenId/Constants.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        public const string AuthenticationErrorUrl = "~/Authentication/Error";
+        public const string LogonCallbackUrl = "~/Users/Account/LogonCallback";
+        public const string OpenIdOwinMiddlewarePriority = "10";
+        public const string LocalIssuer = "LOCAL AUTHORITY";
+        public const string FormsIssuer = "Forms";
+
+        public const string AzureActiveDirectoryObjectIdentifierKey = "http://schemas.microsoft.com/identity/claims/objectidentifier";
+
+        // Certificate Subject Key Identifier
+        public const string VeriSignClass3SecureServerCA_G2 = "A5EF0B11CEC04103A34A659048B21CE0572D7D47";
```
Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.OpenId/Constants.cs]: What the recommendation with updating these strings? When, how?

Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.OpenId/Constants.cs]: I see it seems to be only for Twitter. Could we make these site settings so it can be updated without having to redeploy, or get a new Orchard deployment? It could be an advanced site setting along with the current twitter settings, hidden in a collapsible section as it's just noise. And an admin banner that would be triggered from a date that is also in these settings, based on the first certificate to expire because at some point it will stop working and people will never find out why.

Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.OpenId/Constants.cs]: Remember to cache the values once taken from the settings.

Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.OpenId/Constants.cs]: Is there the same issue for Azure AD, I can remember they regularly recycle some keys or thumbprints.

File: src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+        public void Load() {
+            _sessionLock.EnterReadLock();
+            try
+            {
+                if (InMemoryTokenCache.ContainsKey(_cacheId))
+                    Deserialize(InMemoryTokenCache.Where(x => x.Key == _cacheId).First().Value);
+            }
+            catch (Exception) {
+                _sessionLock.ExitReadLock();
```
Comment (DanielStolt) [inline src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs]: Why not move this to finally block?

File: src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+                HasStateChanged = false;
+
+                if (InMemoryTokenCache.ContainsKey(_cacheId))
+                    InMemoryTokenCache.Remove(_cacheId);
+
+                InMemoryTokenCache.Add(_cacheId, Serialize());
+            }
+            catch (Exception)
+            {
+                _sessionLock.ExitReadLock();
```
Comment (DanielStolt) [inline src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs]: Should probably be `ExitWriteLock()`. (And why not move to finally block?) ;)

File: src/Orchard.Web/Modules/Orchard.OpenId/Module.txt [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+Description: Enables Orchard to authenticate users using OpenId
+Category: OpenId Providers
+Features:
+    Orchard.OpenId:
+        Description: Enables Orchard to authenticate users using OpenId
+		Category: Authentication
+		Name: OpenId
+    Orchard.OpenId.Facebook:
+        Description: Enables Orchard to authenticate users using their Facebook Accounts
+		Category: OpenId Providers
```
Comment (DanielStolt) [inline src/Orchard.Web/Modules/Orchard.OpenId/Module.txt]: I think the category of the providers should be "Authentication" just like the root feature.

File: src/Orchard.Web/Modules/Orchard.OpenId/Constants.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
+﻿namespace Orchard.OpenId {
+    public class Constants {
```
Comment (DanielStolt) [inline src/Orchard.Web/Modules/Orchard.OpenId/Constants.cs]: I think the structure would be nicer if these constants were moved into feature-specific constant files. Since it's a possible extensibility point to write more providers.

**Comment by DanielStolt** [comment]
Truly great contribution! I wrote some code-level comments on your PR.

File: src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs (hunk 4 above)
Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs]: Why not a concurrent dictionary, I think it has the primitives you need (`TryGetOrAdd`)

File: src/Orchard.Web/Modules/Orchard.OpenId/Constants.cs (hunk 3 above)
Comment (ThaerAlAjlouni) [inline src/Orchard.Web/Modules/Orchard.OpenId/Constants.cs]: I couldn't find any formal facts on when those values will expire, but to be in the safe side they are now configurable via admin dashboard as you suggested

File: src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs (hunk 5 above)
Comment (ThaerAlAjlouni) [inline src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs]: No need for locks anymore, the code has been updated to use a Concurrent Dictionary.

File: src/Orchard.Web/Modules/Orchard.OpenId/Module.txt (hunk 6 above)
Comment (ThaerAlAjlouni) [inline src/Orchard.Web/Modules/Orchard.OpenId/Module.txt]: Done :)

File: src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs [hunk 8]
```diff
@@ -22,51 +21,30 @@ public class InMemoryCache : TokenCache {
             Load();
         }
 
-        public static Dictionary<string, byte[]> InMemoryTokenCache {
+        public static ConcurrentDictionary<string, byte[]> InMemoryTokenCache {
```
Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs]: Does it have to be public? If not then remove it and put the initialization in the ctor

File: src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs [hunk 9]
```diff
@@ -1,13 +1,12 @@
...
 using System.Collections.Generic;
 using System.Linq;
-using System.Threading;
 using Microsoft.IdentityModel.Clients.ActiveDirectory;
 
 namespace Orchard.OpenId {
     public class InMemoryCache : TokenCache {
-        private static Dictionary<string, byte[]> _inMemoryTokenCache;
-        private static ReaderWriterLockSlim _sessionLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
+        private static ConcurrentDictionary<string, byte[]> _inMemoryTokenCache;
```
Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs]: Does it have to be static? Or should it be different per tenant?

Comment (ThaerAlAjlouni) [inline src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs]: The keys of the dictionary are unique user ids (the id of the user in the open id provider system, not Orchard's id), so even across multiple tenants the dictionary will be safe.

File: src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs (hunk 8 above)
Comment (ThaerAlAjlouni) [inline src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs]: The InMemoryCache class has been updated according to your comment :)

File: src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs (hunk 9 above)
Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs]: Can you be more specific about the keys? Are they the actual user id or Guids?

Comment (ThaerAlAjlouni) [inline src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs]: Guids generated and assigned to the user by the identity provider. This id will be sent back by the provider after authenticating the user inside the JWT.

File: src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs [hunk 10]
```diff
@@ -0,0 +1,62 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+
+namespace Orchard.OpenId {
+    public class InMemoryCache : TokenCache {
```
Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs]: I looked at the documentation for the token cache. And I think it would be much better if you actually had it registered as a singleton service such that each tenant would have its own list of tokens (not the case right now), and that this list would then be cleared when a tenant is restarted (not the case right now). Also, you would then not need to instantiate a new `InMemoryCache` instance for each request. Also, it needs to be a service such that when we use multiple nodes, we can use a token cache based on Redis for instance. And this service should be tied to the Azure feature.

File: src/Orchard.Web/Modules/Orchard.OpenId/Constants.cs (hunk 7 above)
Comment (ThaerAlAjlouni) [inline src/Orchard.Web/Modules/Orchard.OpenId/Constants.cs]: Done :)

File: src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs (hunk 10 above)
Comment (ThaerAlAjlouni) [inline src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs]: @sebastienros, I am having trouble understanding exactly how to convert this class to a Singleton service, and I couldn't find the documentation of this class, can you please provide me a link.

Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs]: Just have this class implement `ISingletonDependency` and then resolve it instead of instantiating it.

Comment (ThaerAlAjlouni) [inline src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs]: Done

File: src/Orchard.Web/Modules/Orchard.OpenId/Services/AzureActiveDirectory/InMemoryCache.cs [hunk 11]
```diff
@@ -1,39 +1,51 @@
...
-            _cacheId = String.Concat(_userObjectId, CacheIdSuffix);
-
-            if (_inMemoryTokenCache == null)
-                _inMemoryTokenCache = new ConcurrentDictionary<string, byte[]>();
+namespace Orchard.OpenId.Services.AzureActiveDirectory {
+    [OrchardFeature("Orchard.OpenId.AzureActiveDirectory")]
+    public class InMemoryCache : TokenCache, ISingletonDependency {
+        public InMemoryCache() {
+            // On tenant start clear cache.
+            Clear();
```
Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.OpenId/Services/AzureActiveDirectory/InMemoryCache.cs]: Not really necessary as the instance is new on tenant restart. Or there is an issue with the design.

File: src/Orchard.Web/Orchard.Web.csproj [hunk 12]
```diff
@@ -14,7 +14,7 @@
     <AssemblyName>Orchard.Web</AssemblyName>
```
Comment (sebastienros) [inline src/Orchard.Web/Orchard.Web.csproj]: Revert the changes in this file

**Comment by sebastienros** [comment]
Just a few little changes more.

File: src/Orchard.Web/Orchard.Web.csproj (hunk 12 above)
Comment (ThaerAlAjlouni) [inline src/Orchard.Web/Orchard.Web.csproj]: Done

File: src/Orchard.Web/Modules/Orchard.OpenId/Services/AzureActiveDirectory/InMemoryCache.cs (hunk 11 above)
Comment (ThaerAlAjlouni) [inline src/Orchard.Web/Modules/Orchard.OpenId/Services/AzureActiveDirectory/InMemoryCache.cs]: This was an extra step, it is redundant so I will remove it

**Comment by ThaerAlAjlouni** [comment]
Is there anything else to be amended?

**Comment by sebastienros** [comment]
I think it's good, should be merged on thursday
